crypto/dilithium/p_pqdsa_test.cc

```diff
@@ -1108,232 +1108,323 @@ TEST_P(PQDSAParameterTest, KAT) {
+  // Create context of PQDSA type and a key pair
   bssl::UniquePtr<EVP_PKEY_CTX> ctx(EVP_PKEY_CTX_new_id(EVP_PKEY_PQDSA, nullptr));
   ASSERT_TRUE(ctx);
+
+  // Setup the context with specific PQDSA parameters.
+  int nid = GetParam().nid;
   ASSERT_TRUE(EVP_PKEY_CTX_pqdsa_set_params(ctx.get(),nid));
-  ASSERT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
+
+  // Generate a key pair.
   EVP_PKEY *raw = nullptr;
+  ASSERT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
   ASSERT_TRUE(EVP_PKEY_keygen(ctx.get(), &raw));
+  ASSERT_TRUE(raw);
   bssl::UniquePtr<EVP_PKEY> pkey(raw);
 
-  // Extract public key and check it is of the correct size
-  uint8_t *buf = nullptr;
-  size_t buf_size;
-  EXPECT_TRUE(EVP_PKEY_get_raw_public_key(pkey.get(), buf, &buf_size));
-  EXPECT_EQ(pk_len, buf_size);
-
-  buf = (uint8_t *)OPENSSL_malloc(buf_size);
-  ASSERT_NE(buf, nullptr);
-  EXPECT_TRUE(EVP_PKEY_get_raw_public_key(pkey.get(), buf, &buf_size));
-
-  buf_size = 0;
-  EXPECT_FALSE(EVP_PKEY_get_raw_public_key(pkey.get(), buf, &buf_size));
+  // ---- 2. Test key generation with PKEY as a template ----
+  ctx.reset(EVP_PKEY_CTX_new(pkey.get(), nullptr));
+  ASSERT_TRUE(ctx);
 
+  // Generate a key pair.
+  raw = nullptr;
+  ASSERT_TRUE(EVP_PKEY_keygen_init(ctx.get()));
+  ASSERT_TRUE(EVP_PKEY_keygen(ctx.get(), &raw));
+  ASSERT_TRUE(raw);
+  pkey.reset(raw);
+
+  // ---- 3. Test getting raw keys and their size ----
+  size_t pk_len, sk_len;
+
+  // First getting the sizes only.
+  ASSERT_TRUE(EVP_PKEY_get_raw_public_key(pkey.get(), nullptr, &pk_len));
+  ASSERT_TRUE(EVP_PKEY_get_raw_private_key(pkey.get(), nullptr, &sk_len));
+  EXPECT_EQ(pk_len, GetParam().public_key_len);
+  EXPECT_EQ(sk_len, GetParam().private_key_len);
+
+  // Then getting the keys and the sizes.
+  std::vector<uint8_t> pk_raw(pk_len);
+  std::vector<uint8_t> sk_raw(sk_len);
+  ASSERT_TRUE(EVP_PKEY_get_raw_public_key(pkey.get(), pk_raw.data(), &pk_len));
+  ASSERT_TRUE(EVP_PKEY_get_raw_private_key(pkey.get(), sk_raw.data(), &sk_len));
+  EXPECT_EQ(pk_len, GetParam().public_key_len);
+  EXPECT_EQ(sk_len, GetParam().private_key_len);
+
+  // ---- 4. Test failure modes for EVP_PKEY_CTX_pqdsa_set_params. ----
+  // ctx is NULL.
+  ASSERT_FALSE(EVP_PKEY_CTX_pqdsa_set_params(nullptr, nid));
   uint32_t err = ERR_get_error();
   EXPECT_EQ(ERR_LIB_EVP, ERR_GET_LIB(err));
-  EXPECT_EQ(EVP_R_BUFFER_TOO_SMALL, ERR_GET_REASON(err));
-  OPENSSL_free(buf);
-  buf = nullptr;
+  EXPECT_EQ(ERR_R_PASSED_NULL_PARAMETER, ERR_GET_REASON(err));
 
-  // Extract private key and check it is of the correct size
-  EXPECT_TRUE(EVP_PKEY_get_raw_private_key(pkey.get(), buf, &buf_size));
-  EXPECT_EQ((size_t)sk_len, buf_size);
+  // ctx->data is NULL
+  void *tmp = ctx.get()->data;
+  ctx.get()->data = nullptr;
+  ASSERT_FALSE(EVP_PKEY_CTX_pqdsa_set_params(ctx.get(), nid));
+  err = ERR_get_error();
+  EXPECT_EQ(ERR_LIB_EVP, ERR_GET_LIB(err));
+  EXPECT_EQ(ERR_R_PASSED_NULL_PARAMETER, ERR_GET_REASON(err));
```

Comment:
Hmm, we set `ctx->data` to NULL, then pass in `ctx`, expecting a failure because of a null value. This is consistant with how we use the error code in ML-KEM at https://github.com/jakemas/aws-lc/blob/ab8953b7478f72a2b80497cf54ce0b72ed719cfc/crypto/evp_extra/evp_extra_test.cc#L2229

Reply:
It's good to be consistent, I agree. We can later revisit it to see if we want to return another error message. I believe the data would be null if the context wasn't initialised or something.

Reply:
Shall we leave as is for now then? In my opinion the most useful such a generic error message can be is when it is consistant with other failures cases in the library.